crates/iceberg/src/spec/table_metadata_builder.rs

```diff
@@ -0,0 +1,2074 @@
+            .set_properties(properties)
+    }
+
+    /// Creates a new table metadata builder from the given metadata to modify it.
+
+    /// `current_file_location` is the location where the current version
+    /// of the metadata file is stored. This is used to update the metadata log.
+    /// If `current_file_location` is `None`, the metadata log will not be updated.
+    /// This should only be used to stage-create tables.
+    #[must_use]
+    pub fn new_from_metadata(
+        previous: TableMetadata,
+        previous_file_location: Option<String>,
+    ) -> Self {
+        Self {
+            previous_history_entry: previous_file_location.map(|l| MetadataLog {
+                metadata_file: l,
+                timestamp_ms: previous.last_updated_ms,
+            }),
+            metadata: previous,
+            changes: Vec::default(),
+            last_added_schema_id: None,
+            last_added_spec_id: None,
+            last_added_order_id: None,
+        }
+    }
+
+    /// Creates a new table metadata builder from the given table creation.
+    pub fn from_table_creation(table_creation: TableCreation) -> Result<Self> {
+        let TableCreation {
+            name: _,
+            location,
+            schema,
+            partition_spec,
+            sort_order,
+            properties,
+        } = table_creation;
+
+        let location = location.ok_or_else(|| {
+            Error::new(
+                ErrorKind::DataInvalid,
+                "Can't create table without location",
+            )
+        })?;
+        let partition_spec = partition_spec.unwrap_or(UnboundPartitionSpec {
+            spec_id: None,
+            fields: vec![],
+        });
+
+        Self::new(
+            schema,
+            partition_spec,
+            sort_order.unwrap_or(SortOrder::unsorted_order()),
+            location,
+            FormatVersion::V1,
+            properties,
+        )
+    }
+
+    /// Get the current schema with all changes applied up to this point.
+    #[inline]
+    pub fn current_schema(&self) -> &SchemaRef {
+        self.metadata.current_schema()
+    }
+
+    /// Get the current last column id
+    #[inline]
+    pub fn last_column_id(&self) -> i32 {
+        self.metadata.last_column_id
+    }
+
+    /// Get the current last updated timestamp
+    #[inline]
+    pub fn last_updated_ms(&self) -> i64 {
+        self.metadata.last_updated_ms
+    }
+
+    /// Changes uuid of table metadata.
+    pub fn assign_uuid(mut self, uuid: Uuid) -> Self {
+        if self.metadata.table_uuid != uuid {
+            self.metadata.table_uuid = uuid;
+            self.changes.push(TableUpdate::AssignUuid { uuid });
+        }
+
+        self
+    }
+
+    /// Upgrade `FormatVersion`. Downgrades are not allowed.
+    ///
+    /// # Errors
+    /// - Cannot downgrade to older format versions.
+    pub fn upgrade_format_version(mut self, format_version: FormatVersion) -> Result<Self> {
```

Comment:
Hm, `FormatVersion` is an enum already with those two values only.
You mean that if someone upgrades that enum and forgets to update the builder we get an error?

Reply:
Just to ensure that when we add version 3 (which is being finalized) it throws an exception.

Reply:
I added a `match` in https://github.com/apache/iceberg-rust/pull/587/commits/226a485e1d72cdc456b215711015ac830ab3b0c3 so that it won't compile unless someone explicitly adds V3 here.

Reply:
Got it, thanks 👍